### ISSUE TITLE
Integrate haskell-flycheck into haskell-mode

### ISSUE THREAD

**Issue by LukeHoersten** [haskell/haskell-mode#538]
[haskell-flycheck](https://github.com/chrisdone/haskell-flycheck/blob/master/haskell-flycheck.el) it's @chrisdone's [flycheck](http://flycheck.org) implementation that uses a `haskell-process` as the backend.

Flycheck comes with a built-in [flycheck-haskell](https://github.com/flycheck/flycheck-haskell) that uses GHC. The Flycheck author, @lunaryorn, has the policy of not having Flycheck depend on external modes as it's untenable. Since `haskell-flycheck` is dependent on `haskell-process` within `haskell-mode` anyway, I think it makes sense to include @chrisdone's `haskell-flycheck` in this mode.

I'd be open to making the code changes myself and submitting a PR but I'm not sure the best approach to integrating something like this, especially since it's conditional on the availability of Flycheck.

**Comment by lunaryorn** [comment]
This was already discussed somewhere else (can't remember where, currently), and basically I won't merge any of @chrisdone's code unless he agrees personally—that's just a matter of fairness, imho—, and I won't merge it, if I have to maintain it on my own.  It's too complicated for that, and I don't know Haskell Mode well enough, nor do I have the time to get to know it.

IOW, I'll only merge this checker, if @chrisdone agrees, and if he or someone else other than me agrees to maintain this checker as part of this repository.

**Comment by LukeHoersten** [comment]
We discussed in a [haskell-flycheck issue](https://github.com/flycheck/flycheck-haskell/issues/27#issuecomment-72882514) previously.

We don't have to merge it into flychecker. That was just one option. The other option is going into `haskell-mode` which I think we agreed on in our previous discussions. That works nicely because it keeps all the `haskell-process` stuff together.

Either way I want it to turn into something easy enough for people to get that it starts getting wider adoption. That's my whole goal here.

**Comment by lunaryorn** [comment]
@LukeHoersten Oh sorry, there's apparently a misunderstanding.  flycheck-haskell is **not** built-in.  It's a Flycheck extension, which isn't subject to any policies about Flycheck itself.

As a matter of fact, flycheck-haskell already depends on Haskell Mode already, albeit only for some minor stuff.

I understand, that you'd like to have a haskell-process based checker published, but the fact that @chrisdone hasn't published it yet makes me think that it's probably not ready yet for a public release.

Anyhow, I'm not at all opposed to merging everything in flycheck-haskell, as long as @chrisdone agrees, and someone else other than me actually maintains the checker.

**Comment by LukeHoersten** [comment]
I do think it makes sense to preserve the current checker as an option (maybe even the default option) as well. I'm not so concerned with where the package goes, just as long as it's reachable form `package.el`.

**Comment by lunaryorn** [comment]
@LukeHoersten The current checker will stay anyway, since it's part of Flycheck itself, and still has a purpose, e.g. for small haskell sources, which don't have a cabal file.  Flycheck Haskell currently just parses Cabal files and sets up Flycheck accordingly.  It doesn't have any checkers on its own.

But again, the decision whether to make the haskell-process based checker available via package.el is not ours.

**Comment by LukeHoersten** [comment]
- "IOW, I'll only merge this checker, if @chrisdone agrees"
- "Anyhow, I'm not at all opposed to merging everything in flycheck-haskell, as long as @chrisdone agrees,"
- "But again, the decision whether to make the haskell-process based checker available via package.el is not ours."

Be patient and give him a chance to respond. To be clear: I'm not disputing that the decision to merge is up to @chrisdone.

In the mean time, I don't want to confuse the discussion about how we would integrate the checkers with whether @chrisdone is OK with it. @chrisdone has mentioned earlier that integration specifics is his main concern:

"@chrisdone commented on Feb 5
I'm okay with merging this into flycheck-haskell if it unifies a common thing and avoids confusion. I'm concerned about pissing off your users by swapping out the backend with something that requires starting a session and ghci (things that people regularly complain about as being such a chore)."

So this brings me back to my last point: how will this work? We have a few options:
1. Merge `flycheck-haskell` and `haskell-flycheck` into the `flycheck-haskell` package.el then auto-detect `haskell-process` if a session is started *or* just require the user to configure themselves.
2. Merge `haskell-flycheck` into `haskell-mode` to consolidate dependencies and default to the `haskell-process` checker.
3. Keep `haskell-flycheck` and `flycheck-haskell` separate package.el packages and the use of either would auto-enable the corresponding checker.

The focus here is on how does the user specify which checker should be used. If you're using `haskell-process` already, it seems unlikely the user would want to use a separate GHC process for flychecking.

**Comment by lunaryorn** [comment]
@LukeHoersten In my opinion it makes no chance for us to discuss anything as long as he hasn't responded.

**Comment by LukeHoersten** [comment]
From our [previous discussion](https://github.com/flycheck/flycheck-haskell/issues/27#issuecomment-73149580):
> chrisdone commented on Feb 5
> I'm okay with merging this into flycheck-haskell if it unifies a common thing and avoids confusion. I'm concerned about pissing off your users by swapping out the backend with something that requires starting a session and ghci (things that people regularly complain about as being such a chore).

**Comment by lunaryorn** [comment]
@LukeHoersten I read “concerned”, whereas you apparently read “I'm okay” :)

But since @chrisdone agreed, please feel free to open a pull request that adds his Cabal Repl checker to this repository.  If @chrisdone wants, I can give him write access, but I'd like to review the checker before merging it into this package and thus released it to a greater audience.

**Comment by LukeHoersten** [comment]
@lunaryorn, Chris is concerned about changing the default checker but he's OK with merging the packages.

Not changing the default means we need to work out the technical details about how the two checkers will interact and become enabled. Until the technical details are worked out, we will not merge anything. Unless you are ready to contribute to this technical discussion I'm afraid we're stalled.

**Comment by gracjan** [comment]
@lunaryorn: I know that you were in the middle of rewrite and *generalized checker*. How far are you with this?

Currently `haskell-mode` has a lot of ways to check Haskell code, including the new `ide-backend`. There is quite a lot of cleanup needed in `haskell-mode` before we can handle yet one checker.

@LukeHoersten: I knew that `flychecker-haskell` uses `haskell-cabal.el` for a reason but I did not know that it uses `haskell-process.el` for anything. How does it use it?

**Comment by gracjan** [comment]
Note that we still have remnants of `flymake` in the code, those are reported to not always work as in #384 or #130.

**Comment by LukeHoersten** [comment]
@gracjan:

- `haskell-flycheck` is @chrisdone's `haskell-process`-based checker.
- `flycheck-haskell` is @lunaryorn's cabal-integration package for the `ghc-checker` built into `flycheck`. He'd have to provide any more details beyond that.

I don't believe that `haskell-flycheck` depends on anything but `haskell-mode` (for `haskell-process`) and `flycheck`. `haskell-process` is what it uses for cabal integration (via ghci etc).

[My comment above](https://github.com/haskell/haskell-mode/issues/538#issuecomment-84061972) enumerates some of the possibilities and potential tradeoffs of merging `haskell-flycheck` into `haskell-mode` (because of it's `haskell-process` dep) or lumping it in with the `flycheck-haskell` package as just sort of an "extras" package for haskell emacs flycheck stuff.

My personal preference is to merge `haskell-flycheck` into the `flycheck-haskell` package and then just give users a way to switch over the `haskell-process-checker` if they're using `haskell-process` already or something like that (the mechanics still need to be worked out). We'd leave the default as the `ghc-checker` in this case.

**Comment by gracjan** [comment]
I did not know that `haskell-flycheck` is something different than `flycheck-haskell`. It that case at least half of my comment does not apply :)

**Comment by LukeHoersten** [comment]
In hind sight `haskell-flycheck` should really be called `flycheck-haskell-process` and `flycheck-haskell` should be called `flycheck-haskell-cabal` or something like that. `flycheck` proper already has a checker called `ghc-checker`. This ambiguity is exactly what I'm trying to resolve here while simultaneously simplifying configuration and access to these projects.

NOTE: I'm not really proposing those name changes, just using them for illustrative purposes.

**Comment by lunaryorn** [comment]
@gracjan I'm not rewriting the built-in GHC syntax checker.  I wonder where you got this idea from.  I had the idea to use `cabal build` for syntax checking, in this package, but it seems redundant now in light of a haskell-process based checker.

@LukeHoersten We are not stalled then.  I'll contribute to the technical aspects of the syntax checker implementation, with regards to how we can seamlessly switch between GHC and haskell-process based checking depending on whether a running Cabal REPL is available for the current buffer.  Flycheck already provides the necessary infrastructure, and if haskell-process does, too, we shouldn't have any problems.

I'd just like to ask you to open a PR with the new checker first.  It's just easier to discuss technical aspects with the code at hand :)

**Comment by chrisdone** [comment]
To be clear, I wrote haskell-flycheck as an experiment, I'm not convinced it's a good idea, as it has subtle interactions with e.g. the REPL context. Although it seems to work.

Relatedly, [ide-backend](https://github.com/fpco/ide-backend) being open sourced changes the situation. Perhaps in a month we'll have a flycheck checker for [ide-backend-client](https://github.com/chrisdone/ide-backend-client), so this discussion could be for naught. I don't know.

**Comment by gracjan** [comment]
Good to know, thanks for the update.

**Comment by lunaryorn** [comment]
@chrisdone Since you've already implemented it, it's worth a try imho.  If it's too flaky we can always return to the simple GHC checker.  Thanks to Flycheck's automatic checker selection it should be mostly transparent to the user.

I'm not sure whether ide-backend changes the game currently.  If I remember the announcement post aright, it doesn't yet expose compilation errors… if it did, thought, it'd be trivial to call out to `die-backend` to check the current buffer.

**Comment by chrisdone** [comment]
> Since you've already implemented it, it's worth a try imho. If it's too flaky we can always return to the simple GHC checker. Thanks to Flycheck's automatic checker selection it should be mostly transparent to the user.

Fair enough. If it's not disrupting existing users of Flycheck-haskell it seems fine to me.

> I'm not sure whether ide-backend changes the game currently. If I remember the announcement post aright, it doesn't yet expose compilation errors… if it did, thought, it'd be trivial to call out to die-backend to check the current buffer.

It does do that, I just didn't do anything interesting with the output in my emacs minor mode yet. It'll just be a case of calling `getSourceErrors` from the backend and asynchronously waiting on a list of errors, similar to what I'm doing with this haskell-flycheck mode. It should be a trivial checker to write, but I didn't get round to it yet.

**Comment by chrisdone** [comment]
@lunaryorn Example, if interested: https://github.com/chrisdone/ide-backend-mode/blob/master/ide-backend-mode.el#L432 Here I'm just message'ing them but of course that could easily be pushed into a list via flycheck's async API.

**Comment by LukeHoersten** [comment]
Does everyone generally agree that we should be using `ide-backend` as the incremental compile (flycheck), code completion, and code navigation backend?

This means we now have two processes running during dev work: ghci for repl and ide-backend for everything else. This isn't necessarily bad, just clarifying? A goal of mine is to keep the number of external dependencies to a minimum in my dev environment just to keep things simple and fast.

If that's the case, @chrisdone would you just prefer waiting until you've integrated `ide-backend` with flycheck, autocomplete or company, and some code nav system instead of releasing `haskell-flycheck` in some fashion? What's the ETA on the backend integration?

**Comment by chrisdone** [comment]
I don't think anyone yet is using ide-backend, but it's a possible future contender as it's a high-quality implementation worked on by Well-Typed.

> This means we now have two processes running during dev work: ghci for repl and ide-backend for everything else. 

Actually ide-backend supports running statements (REPL stuff) and debugger stuff, although this was never exposed by the public online FP Complete IDE. The functionality is there. So really it would just be "the backend" for everything, ideally speaking. Similar to the `interactive-haskell-mode` minor mode, `ide-backend-mode` provides some keybindings that know about having a "cabal project", loading and queuing commands and such. Others may dislike the fact you have to choose an interaction minor mode, but such infrastructure was anticipating future potential "better" modes.

> If that's the case, @chrisdone would you just prefer waiting until you've integrated ide-backend with flycheck, autocomplete or company, and some code nav system instead of releasing haskell-flycheck in some fashion? What's the ETA on the backend integration?

That's what I'm thinking. Here is the client: https://github.com/chrisdone/ide-backend-client and here is the mode: https://github.com/chrisdone/ide-backend-mode but it's pre-alpha. I don't have an ETA but it's less than an hour's work to get the flycheck checker defined. I don't know company-mode as a library, but it'd probably an hour to integrate autocomplete.el, so assuming the diff between company-mode and autocomplete.el is small then it should also be easy.

The instigator for me will be when I get some spare time I'll start trying to use ide-backend-mode for my real work projects and at that time it'll be a forcing function for fleshing out the rest. I'll probably move both projects under the `haskell` organization at that point.

**Comment by LukeHoersten** [comment]
Makes sense. One of the things I liked about using ghci (or ghc itself) as the backend for flycheck and code nav is that it's kept up to date with ghc by core devs. If FPco decides to move on from ide-backend (hypothetically) then we stop getting support for new lang extensions etc. What does everyone think about this? GHC and GHCi move in lock-step.

**Comment by chrisdone** [comment]
Technically the code nav and span type info come from ghci-ng which is maintained by me (which is why it doesn't support GHC 7.10 yet). GHCi proper is more limited, and harder to contribute to. If FP Complete don't upgrade ide-backend to support 7.12 then someone (e.g. me) will.

**Comment by LukeHoersten** [comment]
Conceivably though if you got your ghci-ng patches into ghci proper it would be carried along with future work?

**Comment by chrisdone** [comment]
It is conceivable.

**Comment by lunaryorn** [comment]
@LukeHoersten `ide-backend` won't go away even if FPComplete stops to support it.  We can still keep developing it in the community.  We need something like it anyway to support editors.  

I'm afraid, though, that you two lost me.  I know nothing about ide-backend beyond what was said in the blog post, and certainly not enough to make a decision about what we should use for Flycheck.

That said, I'll happily merge any checker which you see fit for this repo and keep maintaining as good as I can, and I'll give you any help that you need when it comes to the technical aspects of writing a specific checker for Flycheck.  But I'm afraid I can't contribute anything reasonable to this discussion anymore :)

**Comment by LukeHoersten** [comment]
Since @chrisdone is the primary developer of both `haskell-process` and `ide-backend`, we're going to defer any more work on `flycheck` integrations in the hopes that `ide-backend-mode` supersedes `haskell-process`/`ghci-ng`-based integrations.

**Comment by lunaryorn** [comment]
@LukeHoersten So we'll let this matter rest until ide-backend-mode is ready?

**Comment by LukeHoersten** [comment]
I think so, yes. We'll see how far Chris gets with the implementation.
